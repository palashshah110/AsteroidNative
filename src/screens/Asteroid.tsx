import React, {useState} from 'react';
import {
  SafeAreaView,
  StyleSheet,
  Text,
  View,
  TextInput,
  TouchableOpacity,
  Alert,
  ActivityIndicator,
  ImageBackground,
} from 'react-native';

export default function Asteroid({navigation}) {
  const [AsteroidID, setAsteroidID] = useState<number>(0);
  const [Loading, setLoading] = useState<boolean>(false);
  const handleChange = (Id: any) => {
    if (/^[0-9]*$/.test(Id)) {
      setAsteroidID(Id);
    }
  };

  const handleSubmit = async () => {
    setLoading(true);
    try {
      const response = await fetch(
        `https://api.nasa.gov/neo/rest/v1/neo/${AsteroidID}?api_key=6202jSPpG2GqkXh8LHBaPbumSZ1WVY8evbdOavNs`,
        {method: 'GET'},
      );
      if (response.ok) {
        navigation.navigate('AsteroidDetails', {
          AsteroidData: await response.json(),
        });
      } else {
        Alert.alert('Invalid Asteroid ID');
      }
    } catch (err) {
      Alert.alert('Invalid Asteroid ID');
    } finally {
      setLoading(false);
      setAsteroidID(0);
    }
  };

  const handleRandomClick = async () => {
    setLoading(true);
    try {
      const response = await fetch(
        `https://api.nasa.gov/neo/rest/v1/neo/browse?api_key=6202jSPpG2GqkXh8LHBaPbumSZ1WVY8evbdOavNs`,
        {method: 'GET'},
      );
      const response1 = await response.json();

      const randomID: {id: string}[] = response1?.near_earth_objects.map(
        (item: any) => ({id: item.id}),
      );
      const ranid: number = Math.floor(Math.random() * randomID.length);
      const {id} = randomID[ranid];

      const response2 = await fetch(
        `https://api.nasa.gov/neo/rest/v1/neo/${id}?api_key=6202jSPpG2GqkXh8LHBaPbumSZ1WVY8evbdOavNs`,
        {method: 'GET'},
      );
      if (response2.ok) {
        navigation.navigate('AsteroidDetails', {
          AsteroidData: await response2.json(),
        });
      }
    } catch (err) {
      Alert.alert('Error fetching random asteroid. Please try again.');
    } finally {
      setLoading(false);
      setAsteroidID(0);
    }
  };

  return (
    <ImageBackground
      source={{
        uri: 'https://w0.peakpx.com/wallpaper/107/636/HD-wallpaper-asteroid-espace-space.jpg',
      }}
      resizeMode="cover"
      style={{flex: 1, width: '100%', display: 'flex'}}>
      <SafeAreaView style={styles.container}>
        <View style={styles.inputContainer}>
          <TextInput
            inputMode="tel"
            style={styles.input}
            value={AsteroidID}
            onChangeText={handleChange}
            placeholder="Enter Asteroid ID"
            maxLength={7}
          />
          <View style={styles.btncontainer}>
            <TouchableOpacity
              disabled={Loading || AsteroidID.toString().length < 7}
              style={
                Loading || AsteroidID.toString().length < 7
                  ? styles.submitDisButton
                  : styles.submitButton
              }
              onPress={handleSubmit}
              testID="submitbtn">
              {Loading ? (
                <View
                  style={{
                    flex: 1,
                    justifyContent: 'center',
                    alignItems: 'center',
                  }}>
                  <ActivityIndicator size="large" testID="loading-indicator" />
                </View>
              ) : (
                <Text style={styles.submitButtonText}>Search</Text>
              )}
            </TouchableOpacity>
            <TouchableOpacity
              style={styles.submitButton}
              onPress={handleRandomClick}
              testID="randombtn">
              <Text style={styles.submitButtonText}>Random</Text>
            </TouchableOpacity>
          </View>
        </View>
      </SafeAreaView>
    </ImageBackground>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  inputContainer: {
    flexDirection: 'column',
    alignItems: 'center',
    height: 100,
  },
  input: {
    backgroundColor: '#fff',
    borderWidth: 1,
    borderColor: 'black',
    padding: 15,
    borderRadius: 10,
    flex: 1,
    width: 250,
  },
  submitButton: {
    backgroundColor: 'blue',
    padding: 10,
    borderRadius: 5,
    marginTop: 10,
    width: 100,
  },
  submitDisButton: {
    backgroundColor: 'blue',
    padding: 10,
    borderRadius: 5,
    marginTop: 10,
    width: 100,
    opacity: 0.2,
  },
  submitButtonText: {
    color: 'white',
    fontWeight: 'bold',
    textAlign: 'center',
  },
  btncontainer: {
    flexDirection: 'row',
    gap: 10,
    height: 50,
  },
});
