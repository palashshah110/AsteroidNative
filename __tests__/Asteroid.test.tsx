import React, {act} from 'react';
import Asteroid from '../src/screens/Asteroid';
import {
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react-native';
import fetchMock from 'jest-fetch-mock';
import {Alert} from 'react-native';
fetchMock.enableMocks();

const mockData1: {id: number; name: string}[] = [
  {id: 3542519, name: 'as'},
  {id: 3542519, name: 'asp'},
];
const mockData2: {id: number}[] = [{id: 3542519}, {id: 3542519}];

const mockNavigation = {navigate: jest.fn()};

describe('Asteroid Form Component', () => {
  afterEach(() => {
    fetchMock.resetMocks();
  });
  test('Checking id and submit Button', () => {
    render(<Asteroid navigation={mockNavigation} />);
    const AsteroidInput = screen.getByPlaceholderText('Enter Asteroid ID');
    fireEvent.changeText(AsteroidInput, 3542519);
    expect(AsteroidInput.props.value).toBe(3542519);
    const AsteroidBtn = screen.getByTestId('submitbtn');
    fireEvent.press(AsteroidBtn);
    expect(fetchMock).toHaveBeenCalledTimes(1);
  });

  test('Checking Random Btn', () => {
    render(<Asteroid navigation={mockNavigation} />);
    const randombtn = screen.getByTestId('randombtn');
    fireEvent.press(randombtn);

    expect(fetchMock).toHaveBeenCalledTimes(1);

    fetchMock.mockReturnValue(mockData2);
    const randomID: {id: string}[] = mockData2.map((item: any) => ({
      id: item.id,
    }));
    const ranid: number = Math.floor(Math.random() * randomID.length);
    const {id} = randomID[ranid];
    expect(fetchMock).toHaveBeenCalledTimes(1);
    // expect(fetchMock).toHaveBeenCalledWith(
    //   `https://api.nasa.gov/neo/rest/v1/neo/${id}?api_key=6202jSPpG2GqkXh8LHBaPbumSZ1WVY8evbdOavNs`,
    // );
  });

  // test('Checking Random Btn', async () => {
  //   render(<Asteroid navigation={mockNavigation} />);
  //   const randombtn = screen.getByTestId('randombtn');
  //   fireEvent.press(randombtn);
  // await waitFor(() => expect(fetchMock).toHaveBeenCalledTimes(2));
  // expect(mockNavigation.navigate).toHaveBeenCalledWith(
  //   'AsteroidDetails',
  //   expect.any(Object),
  // );
  // });

  test('should show an error when API  fetchdata give error', async () => {
    jest.spyOn(Alert, 'alert');
    render(<Asteroid navigation={{navigate: jest.fn()}} />);

    fetchMock.mockRejectOnce(() => Promise.reject('API error'));

    const AsteroidInput = screen.getByPlaceholderText('Enter Asteroid ID');
    fireEvent.changeText(AsteroidInput, 12345678);
    expect(AsteroidInput.props.value).toBe(12345678);

    const AsteroidBtn = screen.getByTestId('submitbtn');
    fireEvent.press(AsteroidBtn);

    expect(fetchMock).toHaveBeenCalledTimes(1);

    expect('Invalid Asteroid ID').toMatchSnapshot();
  });

  test('should show an error when API  Random Btn give error', async () => {
    jest.spyOn(Alert, 'alert');
    render(<Asteroid navigation={{navigate: jest.fn()}} />);

    fetchMock.mockRejectOnce(() => Promise.reject('API error'));

    const randombtn = screen.getByTestId('randombtn');
    fireEvent.press(randombtn);

    expect(fetchMock).toHaveBeenCalledTimes(1);

    expect(
      'Error fetching random asteroid. Please try again.',
    ).toMatchSnapshot();
  });
});
