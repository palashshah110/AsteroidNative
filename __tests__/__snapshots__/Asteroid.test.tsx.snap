// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asteroid Form Component should show an error when API  Random Btn give error 1`] = `"Error fetching random asteroid. Please try again."`;

exports[`Asteroid Form Component should show an error when API  fetchdata give error 1`] = `"Invalid Asteroid ID"`;
